pkg/solana/txm/pendingtx.go

```diff
@@ -556,6 +567,129 @@ func (c *pendingTxContext) TrimFinalizedErroredTxs() int {
 	return len(expiredIDs)
 }
 
+func (c *pendingTxContext) GetSignatureInfo(sig solana.Signature) (txInfo, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	info, exists := c.sigToTxInfo[sig]
+	if !exists {
+		return txInfo{}, ErrSigDoesNotExist
+	}
+	return info, nil
+}
+
+func (c *pendingTxContext) OnReorg(sig solana.Signature) (pendingTx, error) {
+	// Acquire a read lock to check if the signature exists and needs to be reset
+	err := c.withReadLock(func() error {
+		// Check if the signature is still being tracked
+		info, exists := c.sigToTxInfo[sig]
+		if !exists {
+			return ErrSigDoesNotExist
+		}
+
+		// Check if the transaction is still in a non finalized/errored state
+		var broadcastedExists, confirmedExists bool
+		_, broadcastedExists = c.broadcastedProcessedTxs[info.id]
+		_, confirmedExists = c.confirmedTxs[info.id]
+		if !broadcastedExists && !confirmedExists {
+			return ErrTransactionNotFound
+		}
+		return nil
+	})
+	if err != nil {
+		// If transaction or sig are not found, return
+		return pendingTx{}, err
+	}
+
+	var pTx pendingTx
+	// Acquire a write lock to perform the state reset
+	_, err = c.withWriteLock(func() (string, error) {
+		// Retrieve sig and tx again inside the write lock
+		info, exists := c.sigToTxInfo[sig]
+		if !exists {
+			return "", ErrSigDoesNotExist
+		}
+
+		tx, broadcastedProcessedExists := c.broadcastedProcessedTxs[info.id]
+		if broadcastedProcessedExists {
+			pTx = tx
+		}
+		tx, confirmedExists := c.confirmedTxs[info.id]
+		if confirmedExists {
+			pTx = tx
+		}
+
+		if !broadcastedProcessedExists && !confirmedExists {
+			// transaction does not exist in any non finalized/errored maps
+			return "", ErrTransactionNotFound
+		}
+
+		// reset state to broadcasted and update the transaction in the broadcasted map
+		info.state, pTx.state = Broadcasted, Broadcasted
+		c.sigToTxInfo[sig] = info
+		c.broadcastedProcessedTxs[info.id] = pTx
+
+		// If the transaction regressed from confirmed state, we also need to remove it from the confirmed map
+		if confirmedExists {
+			delete(c.confirmedTxs, info.id)
+		}
+
+		return "", nil
+	})
+	if err != nil {
+		// If transaction or sig were not found
+		return pendingTx{}, err
+	}
+
+	// Returns the transaction in case we need to rebroadcast and restart the retry/bumping cycle
+	return pTx, nil
+}
+
+// TxHasReorg determines whether a reorg has occurred for a given tx.
+// It achieves this by comparing the highest aggregated state across all associated signatures with the current state of the transaction.
+// If the highest aggregated state is less than the current state, a reorg has occurred and we need to handle it.
+func (c *pendingTxContext) TxHasReorg(id string) bool {
+	var pTx pendingTx
+	var broadcastedExists, confirmedExists bool
+	statePriority := map[TxState]int{
+		Broadcasted: 1,
+		Processed:   2,
+		Confirmed:   3,
+	}
```

Comment:
Not against this if it helps but I think the `TxState` enum is ordered in this way already. Couldn't we just compare these states against each other without the map?

Reply:
yes! this makes things simpler